### PR TITLE
Only save artifacts when they are directly referenced

### PR DESCRIPTION
**Pull request by alexcb**
Re https://github.com/earthly/earthly/issues/896

Artifacts will only be saved locally when their target is directly
referenced, or referenced by a BUILD.

This will cause targets that are referenced from a FROM, or COPY from
*not* saving artifacts as a side-effect.

Signed-off-by: Alex Couture-Beil <alex@earthly.dev>

### CODE REVIEW

**Comment by vladaionescu** [comment]
Remember to feature-flag this ;-)

File: earthfile2llb/converter.go [hunk 1]
```diff
@@ -1122,7 +1122,7 @@ func (c *Converter) ExpandArgs(word string) string {
 	return c.varCollection.Expand(word)
 }
 
-func (c *Converter) prepBuildTarget(ctx context.Context, fullTargetName string, platform *specs.Platform, allowPrivileged bool, buildArgs []string, isDangling bool, isFrom bool) (domain.Target, ConvertOpt, bool, error) {
+func (c *Converter) prepBuildTarget(ctx context.Context, fullTargetName string, platform *specs.Platform, allowPrivileged, doSaves bool, buildArgs []string, isDangling bool, isFrom bool) (domain.Target, ConvertOpt, bool, error) {
```
Comment (vladaionescu) [inline earthfile2llb/converter.go:1125]: I would merge `doSaves` and `isFrom` into a single parameter `cmdName` and use that.

```go
opt.doSaves = (cmdName == 'BUILD' && c.opt.doSaves)
```

**Review by vladaionescu — COMMENTED** [review]
I feel like the logic is split too much between the interpreter and the converter making it hard to follow.

I think you can keep everything only in converter and simply check in `buildTarget` / `prepBuildTarget` whether the command is `BUILD` or not, in order to pass the `opt.doSaves` onwards.

Then the decision to not do anything on saves can be done in the `converter.SaveArtifact` and`converter.SaveImage`.

This should keep all the logic in the converter and make the interpreter decoupled.

File: earthfile2llb/converter.go (hunk 1 above)
Comment (alexcb) [reply earthfile2llb/converter.go]: good plan, that made it much cleaner.

**Comment by vladaionescu** [comment]
This is still being worked on, right? This still needs the changes in `builder.go`.

File: earthfile2llb/earthfile2llb.go [hunk 2]
```diff
@@ -105,18 +111,19 @@ func Earthfile2LLB(ctx context.Context, target domain.Target, opt ConvertOpt) (m
...
 	err = features.ApplyFlagOverrides(ftrs, opt.FeatureFlagOverrides)
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to apply version feature overrides")
 	}
+	opt.Features = ftrs
 
-	if ftrs.ReferencedSaveOnly {
-		fmt.Printf("TODO feature-flip referenced save artifact as local feature in a future PR.\n")
+	if !ftrs.ReferencedSaveOnly {
+		// 0.5 behaviour is to save all artifacts and images in all cases (if they are directly referenced, or indirectly triggered)
```
Comment (vladaionescu) [inline earthfile2llb/earthfile2llb.go:125]: Not really.. in 0.5 artifacts are to be saved if they are part of local targets, whereas images are always saved.

File: builder/builder.go [hunk 3]
```diff
@@ -511,46 +526,44 @@ func (b *Builder) convertAndBuild(ctx context.Context, target domain.Target, opt
...
-					artifact := domain.Artifact{
-						Target:   sts.Target,
-						Artifact: saveLocal.ArtifactPath,
-					}
-					err := b.saveArtifactLocally(ctx, artifact, artifactDir, saveLocal.DestPath, sts.ID, opt, saveLocal.IfExists)
-					if err != nil {
-						return nil, err
-					}
-					dirIndex++
+			for _, saveLocal := range sts.SaveLocals {
```
Comment (vladaionescu) [inline builder/builder.go:529]: Shouldn't the `!sts.Target.IsRemote()` still be checked somewhere for 0.5 behavior?

File: earthfile2llb/earthfile2llb.go (hunk 2 above)
Comment (alexcb) [reply earthfile2llb/earthfile2llb.go:125]: Good catch, I will expand the tests to make sure the save images still occur.

File: builder/builder.go (hunk 3 above)
Comment (alexcb) [reply builder/builder.go:529]: I needed to remove it to support the case of performing a build such as:

    earthly github.com/earthly/earthly:main+deps

which would then save the `go.mod` and `go.sum` files locally; however if the target is remote, we perform this extra check earlier up:

```
	if target.IsRemote() {
		if len(featureFlagOverrides) != 0 {
			featureFlagOverrides += ","
		}
		featureFlagOverrides += "referenced-save-only"
	}
```

which will prevent the remote target from saving anything else that was indirectly referenced.

I was thinking that in this case it's fair to assume that if the user is directly referencing a remote target, then that user wants any directly referenced files to be saved.

There might be an issue where a user gives us a local Earthfile that then references something that's remote, and none of these Earthfiles have any VERSION flags, then those remote targets might also be saved -- let me double check this case (and write a test for it if one doesn't exist).

File: earthfile2llb/earthfile2llb.go (hunk 2 above)
Comment (alexcb) [reply earthfile2llb/earthfile2llb.go:125]: @vladaionescu I just looked over https://github.com/earthly/earthly/issues/896 again and it says:

> We change the logic s.t. output is only produced for the targets invoked either directly (mentioned directly on the command-line) or via calls to BUILD. FROM and COPY (and others e.g. WITH DOCKER --load) do not cause images to be output locally.

Is this still correct? I have this code only locally saving artifacts or images when they are referenced via `FROM` or `BUILD`.

File: builder/builder.go (hunk 3 above)
Comment (alexcb) [reply builder/builder.go:528]: In addition this also had to be added to https://github.com/earthly/earthly/blob/selective-push/earthfile2llb/converter.go#L1199

**Review by vladaionescu — APPROVED** [review]
